xdgopenproxy/xdgopenproxy.go

```diff
@@ -27,37 +27,95 @@ import (
 	"github.com/godbus/dbus"
 )
 
-func Run(urlOrFile string) error {
-	bus, err := dbus.SessionBus()
+type bus interface {
+	Object(name string, objectPath dbus.ObjectPath) dbus.BusObject
+	Close() error
+}
+
+var sessionBus = func() (bus, error) { return dbus.SessionBus() }
+
+type desktopLauncher interface {
+	openFile(path string) error
+	openURL(url string) error
+}
+
+type portalLauncher struct {
+	service dbus.BusObject
+}
+
+func (p *portalLauncher) openFile(filename string) error {
+	fd, err := syscall.Open(filename, syscall.O_RDONLY, 0)
 	if err != nil {
 		return err
 	}
-	defer bus.Close()
-	launcher := bus.Object("io.snapcraft.Launcher", "/io/snapcraft/Launcher")
-	return launch(launcher, urlOrFile)
+	defer syscall.Close(fd)
+
+	return p.service.Call("org.freedesktop.portal.OpenURI.OpenFile", 0, "", dbus.UnixFD(fd),
+		map[string]dbus.Variant{}).Err
```

Comment:
One other issue: xdg-desktop-portal reports its errors asynchronously, so we don't actually know if the `OpenFile` request was successful at this point.  We need to do something like:

1. add a match rule for `type='signal',sender='org.freedesktop.portal.Desktop`,interface='org.freedesktop.portal.Request',member=`Response'`, and spawn a goroutine to watch signal messages.
2. extract the handle from the method call's return value.
3. when we see a signal emission with the matching object path, check the first argument of the signal to see if the call was successful.